### PR TITLE
Adding invalid split messages

### PR DESCRIPTION
**Pull request by Rambatino**
Added invalid split messages to Split, such that it's obvious why it terminated there.

### CODE REVIEW

File: CHAID/tree.py [hunk 1]
```diff
@@ -159,13 +168,9 @@ def to_tree(self):
 
     def __iter__(self):
         """ Function to allow nodes to be iterated over """
-        if not self.tree_store:
-            self.build_tree()
         return iter(self.tree_store)
```
Comment (Rambatino) [inline CHAID/tree.py]: This is so much cleaner IMO

File: CHAID/stats.py [hunk 2]
```diff
@@ -90,9 +94,16 @@ def best_cat_split(self, ind, dep):
                     if choice is None or p_split > highest_p_join or (p_split == highest_p_join and chi > split_chi):
                         choice, highest_p_join, split_chi = comb, p_split, chi
 
-                sufficient_split = highest_p_join < self.alpha_merge and all(
+                invalid_key = None
+                sufficient_split = highest_p_join < self.alpha_merge
+                if not sufficient_split: invalid_key = 'alpha_merge'
+
+                sufficient_split = sufficient_split and all(
+                    # what if a greater p-value on a different grouping would satisfy alpha merge _and_ min_child_node_size?
```
Comment (Rambatino) [inline CHAID/stats.py]: @xulaus could you put your brain on that question...

**Comment by justfish09** [comment]
👍

File: CHAID/invalid.py [hunk 3]
```diff
@@ -0,0 +1,12 @@
+class Invalid:
```
Comment (Rambatino) [inline CHAID/invalid.py]: Could do, how do you want to proceed?

File: CHAID/split.py [hunk 4]
```diff
@@ -16,8 +16,10 @@ class Split(object):
         The p value of that split
     dof : int
         The degrees of freedom as a result of this split
+    invalid_reason : str
```
Comment (xulaus) [inline CHAID/split.py]: Not a str no more

File: tests/test_split.py [hunk 5]
```diff
@@ -16,3 +16,16 @@ def test_column_name_mappings():
         split.name_columns(column_names)
         assert split.column == column_names[split.column_id], 'Names should correctly map to column name when mapping is applied'
 
+def test_invalid_reason_can_be_setted_and_getted():
+    """ Test that invalid reasons can be set and accessed on a split instance """
+    split = CHAID.Split("a", None, None, 1, 0)
+    reason = "test reason"
+    split.invalid_reason = reason
```
Comment (xulaus) [inline tests/test_split.py]: This should be a real reason now, not just a string

**Pull request merged**